roadmap/implementers-guide/src/node/approval/approval-voting.md

```diff
@@ -198,27 +225,71 @@ On receiving an `ApprovedAncestor(Hash, BlockNumber, response_channel)`:
-  * Ensure the validator index is not part of the backing group for the candidate.
-  * Ensure the validator index is not present in the approval entry already.
-  * Create a tranche entry for the delay tranche in the approval entry and note the assignment within it.
-  * Note the candidate index within the approval entry.
-  * Schedule a wakeup with `next_wakeup`.
+#### Schedule Wakeup
+  * Requires `(approval_entry, candidate_entry)` which effectively denotes a `(Block Hash, Candidate Hash)` pair - the candidate, along with the block it appears in.
+  * If the `approval_entry` is approved, this doesn't need to be woken up again.
+  * Return the earlier of our next no-show timeout or the tranche of our assignment, if not yet triggered
+  * Our next no-show timeout is computed by finding the earliest-received assignment within `n_tranches` for which we have not received an approval and adding `to_ticks(session_info.no_show_slots)` to it.
+  * If the approval entry is already approved, or we have triggered our assignment and there are no pending no-shows, we do not need to schedule a wakeup. Note that the latter case is only possible when we have not seen enough assignments in order to approve. When we receive an incoming assignment, we will schedule a new wakeup, and the `(Block, Candidate)` pair will continue to be processed appropriately.
 
-#### `import_checked_approval(BlockEntry, CandidateEntry, ValidatorIndex)`
-  * Set the corresponding bit of the `approvals` bitfield in the `CandidateEntry` to `1`. If already `1`, return.
-  * For each `ApprovalEntry` in the `CandidateEntry` (typically only 1), check whether the validator is assigned as a checker.
-    * If so, set `n_tranches = tranches_to_approve(approval_entry, tranche_now(block.slot, now()))`.
-    * If `check_approval(block_entry, approval_entry, n_tranches)` is true, set the corresponding bit in the `block_entry.approved_bitfield`.
+#### Launch Approval Work
+  * Requires `(SessionIndex, SessionInfo, CandidateReceipt, ValidatorIndex, block_hash, candidate_index)`
+  * Extract the public key of the `ValidatorIndex` from the `SessionInfo` for the session.
+  * Issue an `AvailabilityRecoveryMessage::RecoverAvailableData(candidate, session_index, response_sender)`
+  * Load the historical validation code of the parachain by dispatching a `RuntimeApiRequest::HistoricalValidationCode(`descriptor.para_id`, `descriptor.relay_parent`)` against the state of `block_hash`.
+  * Spawn a background task with a clone of `background_tx`
+    * Wait for the available data
+    * Issue a `CandidateValidationMessage::ValidateFromExhaustive` message
+    * Wait for the result of validation
+    * If valid, issue a message on `background_tx` detailing the request.
 
-#### `tranches_to_approve(approval_entry, tranche_now) -> RequiredTranches`
+#### Issue Approval Vote
+  * Fetch the block entry and candidate entry. Ignore if `None` - we've probably just lost a race with finality.
```

Comment:
We're fetching these from the current chain state?  I suppose no they were purged from our local state, due to finality, yes?

Reply:
Block finality triggers a DB cleanup. whereas it doesn't actually clean up the scheduled wakeups that we have. we just lazily handle wakeups for stuff that's been deleted (i.e. finalized) by ignoring them here.